### PR TITLE
add TensorProxy.__setitem__

### PR DESCRIPTION
**Pull request by t-vi**
This adds `__setitem__` by creating an out of place `thunder.torch.setitem` and inplace `thunder.torch.setitem_`.
 
Fixes: #341 .

### CODE REVIEW

File: thunder/core/functionalization.py [hunk 1]
```diff
@@ -292,10 +292,16 @@ def create_functional_bsym_from(inplace_bsym: BoundSymbol) -> BoundSymbol:
         flat_args, flat_args_spec = tree_flatten((args, kwargs))
         flat_args[optional_inplace_arg_index] = False
         args, kwargs = tree_unflatten(flat_args, flat_args_spec)
+    functional_output = inplace_bsym.output
+    if inplace_bsym.sym is setitem_:
+        # setitem does not return a value, take the output of the setitem subsymbol
+        assert inplace_bsym.subsymbols[0].sym is setitem
+        functional_output = inplace_bsym.subsymbols[0].output
+    print("########", inplace_bsym)
```
Comment (lantiga) [inline thunder/core/functionalization.py:300]: remove

File: thunder/core/prims.py [hunk 2]
```diff
@@ -3380,6 +3381,14 @@ def take_along_axis_meta(a: TensorProxy, /, index: TensorProxy, dim: int) -> Ten
 take_along_axis = make_prim(PrimIDs.TAKE_ALONG_AXIS, "take_along_axis", meta=take_along_axis_meta)
 
 
+def copy_with_setitem_meta(a: TensorProxy, index, value: TensorProxy) -> TensorProxy:
+    # TOO: port checks from clang, currently there  because of the utilities they need
```
Comment (lantiga) [inline thunder/core/prims.py:3385]:
```suggestion
    # TODO: port checks from clang, currently there  because of the utilities they need
```

**Review by lantiga — APPROVED** [review]
👍

File: thunder/core/prims.py (hunk 2 above)
Comment (t-vi) [reply thunder/core/prims.py]: fixed, thank you

**Comment by crcrpar** [comment]
@t-vi would multidimensional indexing work?

**Pull request merged**